### ISSUE TITLE
`ModelMethodOrder` output isn't helpful

### ISSUE THREAD

**Issue by Gummibeer** [tightenco/tlint#162]
The output provided by the `ModelMethodOrder` lint isn't helpful. Below is my `User` model which the linter complains about but so far I got the method order is right!?

* relationships: there are none
* scopes: there are none
* accessors: first two method `getFullNameAttribute` and `getAvatarAttribute`
* mutators: third method `setPasswordAttribute`
* boot: there isn't one
* everything else is placed at the end

I would expect the lint the provide the order it expects the methods to be. 

```bash
$ tlint lint app

Lints for /Users/tomwitkowski/git/Gummibeer/shootager/app/Models/User.php
============
! Model method order should be relationships > scopes > accessors > mutators > boot
55 : `class User extends Model implements AuthenticatableContract, AuthorizableContract, CanResetPasswordContract, MustVerifyEmailContract, GetsCleanedUp`
```

<details>
  <summary>app/Models/User.php</summary>

```php
<?php

namespace App\Models;

use App\Notifications\ResetPassword as ResetPasswordNotification;
use App\Notifications\VerifyEmail as VerifyEmailNotification;
use Astrotomic\Unavatar\Unavatar;
use Illuminate\Auth\Authenticatable;
use Illuminate\Auth\MustVerifyEmail;
use Illuminate\Auth\Passwords\CanResetPassword;
use Illuminate\Contracts\Auth\Access\Authorizable as AuthorizableContract;
use Illuminate\Contracts\Auth\Authenticatable as AuthenticatableContract;
use Illuminate\Contracts\Auth\CanResetPassword as CanResetPasswordContract;
use Illuminate\Contracts\Auth\MustVerifyEmail as MustVerifyEmailContract;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Foundation\Auth\Access\Authorizable;
use Illuminate\Notifications\DatabaseNotification;
use Illuminate\Notifications\DatabaseNotificationCollection;
use Illuminate\Notifications\Notifiable;
use Illuminate\Support\Carbon;
use Illuminate\Support\Facades\Hash;
use Spatie\ModelCleanup\CleanupConfig;
use Spatie\ModelCleanup\GetsCleanedUp;

/**
 * App\Models\User.
 *
 * @property int $id
 * @property string $first_name
 * @property string $last_name
 * @property string|null $business_name
 * @property string $email
 * @property string $password
 * @property string|null $remember_token
 * @property Carbon|null $email_verified_at
 * @property Carbon|null $created_at
 * @property Carbon|null $updated_at
 * @property-read string $full_name
 * @property-read string $avatar
 * @property-read DatabaseNotificationCollection|DatabaseNotification[] $notifications
 * @property-read int|null $notifications_count
 *
 * @method static Builder|User newModelQuery()
 * @method static Builder|User newQuery()
 * @method static Builder|User query()
 * @method static Builder|User whereBusinessName($value)
 * @method static Builder|User whereCreatedAt($value)
 * @method static Builder|User whereEmail($value)
 * @method static Builder|User whereEmailVerifiedAt($value)
 * @method static Builder|User whereFirstName($value)
 * @method static Builder|User whereId($value)
 * @method static Builder|User whereLastName($value)
 * @method static Builder|User wherePassword($value)
 * @method static Builder|User whereRememberToken($value)
 * @method static Builder|User whereUpdatedAt($value)
 * @mixin \Eloquent
 */
class User extends Model implements AuthenticatableContract, AuthorizableContract, CanResetPasswordContract, MustVerifyEmailContract, GetsCleanedUp
{
    use Notifiable, Authenticatable, Authorizable, CanResetPassword, MustVerifyEmail;

    protected $fillable = [
        'first_name',
        'last_name',
        'business_name',
        'email',
        'password',
    ];

    protected $hidden = [
        'password',
        'remember_token',
    ];

    protected $dates = [
        'email_verified_at',
    ];

    protected $casts = [];

    public function getFullNameAttribute(): string
    {
        return sprintf('%s %s', $this->first_name, $this->last_name);
    }

    public function getAvatarAttribute(): string
    {
        return Unavatar::email($this->email)->toUrl();
    }

    public function setPasswordAttribute(string $value): void
    {
        if (Hash::info($value)['algo'] === null) {
            $value = Hash::make($value);
        }

        $this->attributes['password'] = $value;
    }

    public function cleanUp(CleanupConfig $config): void
    {
        $config
            ->olderThanDays(7)
            ->scope(fn (Builder $query) => $query->whereNull('email_verified_at'));
    }

    public function sendEmailVerificationNotification(): void
    {
        $this->notify(new VerifyEmailNotification($this));
    }

    public function sendPasswordResetNotification($token): void
    {
        $this->notify(new ResetPasswordNotification($this, $token));
    }
}
```

</details>

**Comment by Gummibeer** [comment]
Okay, seems like I've found the issue.
```php
/** This is the catch all, as custom methods should be extracted into "service objects" */
return 'relationship';
```

So all custom methods are detected as `relationship`.

In this case the linter does more than it's expected to do - it should **only** validate the method order but not the method types.
Could we change this to `custom` and add them to the end of the method order and optionally add a new linter that validates that the model does not contain additional methods?

The problem is that even Laravel itself defines a ton of additional methods on the model and almost every package expects methods on the model if they interact with one.
So preventing additional methods - or forcing them to the top of the file feels wrong.

**Comment by loganhenson** [comment]
> Could we change this to custom and add them to the end of the method order and optionally add a new linter that validates that the model does not contain additional methods

This makes total sense to me 👌

**Comment by Gummibeer** [comment]
I'm starting to work on this.
**Plus** `booting` and `booted` method types - which are the new preferred way of doing model boot stuff because you don't have to call `parent::...()`.

Should I add the boot things in its own PR?

**Comment by Gummibeer** [comment]
Okay, I'm doing it altogether because I'm also reworking the `ModelMethodOrder` description output to tell the user what's expected.

**Comment by Gummibeer** [comment]
The linter description output is now something like:
```
Model method order should be: relationships > scopes > accessors > mutators > booting > boot > booted > custom
Methods are expected to be ordered like:
 * category() is matched as "relationship"
 * phone() is matched as "relationship"
 * tags() is matched as "relationship"
 * scopeWhereInactive() is matched as "scope"
 * scopeWhereActive() is matched as "scope"
 * getFirstNameAttribute() is matched as "accessor"
 * setFirstNameAttribute() is matched as "mutator"
 * setLastNameAttribute() is matched as "mutator"
 * booting() is matched as "booting"
 * boot() is matched as "boot"
 * booted() is matched as "booted"
 * activate() is matched as "custom"
```

So a user can exactly compare and also directly sees how his method is matched. If he believes that a method is matched with the wrong type he can check if the method misses a return type or has a typo otherwise raise an issue.

Because `relationship` methods don't have any special naming rule they are detected by return type. So every relationship method without an explicit `: HasOne` or similar is matched as `custom`.
This could lead to a breaking change.
